In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import numpy as np

###Comparativa de DataFrames:

In [3]:
class display(object):
    """Display HTML representation of multiple objects"""
    template = """<div style="float: left; padding: 10px;">
    <p style='font-family:"Courier New", Courier, monospace'>{0}</p>{1}
    </div>"""
    def __init__(self, *args):
        self.args = args
        
    def _repr_html_(self):
        return '\n'.join(self.template.format(a, eval(a)._repr_html_())
                         for a in self.args)
    
    def __repr__(self):
        return '\n\n'.join(a + '\n' + repr(eval(a))
                           for a in self.args)

In [4]:
df_W = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data Science - Coder/Archivos csv/w_data.csv', sep=';')
df_P = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data Science - Coder/Archivos csv/p_data.csv', sep=';')

###Leemos ambos CSV y los mostramos juntos, cada uno en su DataFrame:

In [ ]:
display('df_P', 'df_W')

df_P
        Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  
0       Bachelors           13.0        Never-married    Not-in-family  
1       Bachelors           13.0   Married-civ-spouse          Husband  
2         HS-grad            9.0             Divorced    Not-in-family  
3            11th            7.0   Married-civ-spouse          Husband  
4       Bachelors           13.0   Married-civ-spouse             Wife  
...           ...            ...                  ...              ...  
48837   Bachelors           13.0        Never-married        Own-child  
48838   Bachelors           13.0             Divorced    Not-in-family  
48839     HS-grad            9.0              Widowed   Other-relative  
48840   Bachelors           13.0   Married-civ-spouse          Husband  
48841   Bachelors           13.0             Divorced        Own-child  

[48842 rows x 9 columns]

df_W
               Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target  
0                40.0       49084    <=50K  
1                13.0       68553    <=50K  
2                40.0      143654    <=50K  
3                40.0       41510    <=50K  
4                40.0      195025    <=50K  
...               ...         ...      ...  
48838            36.0   735617153   <=50K.  
48839            40.0  1131807446   <=50K.  
48840            50.0   781422087   <=50K.  
48841            40.0   161771296   <=50K.  
48842            60.0  1344995623    >50K.  

[48843 rows x 7 columns]

###Armamos el DataFrame agrupado ("Total"), fusionando ambos a partir de la variable "ID":

In [ ]:
df_Total = pd.merge(df_P,df_W, on = 'ID',how='outer')

display('df_P', 'df_W','df_Total')

df_P
        Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  
0       Bachelors           13.0        Never-married    Not-in-family  
1       Bachelors           13.0   Married-civ-spouse          Husband  
2         HS-grad            9.0             Divorced    Not-in-family  
3            11th            7.0   Married-civ-spouse          Husband  
4       Bachelors           13.0   Married-civ-spouse             Wife  
...           ...            ...                  ...              ...  
48837   Bachelors           13.0        Never-married        Own-child  
48838   Bachelors           13.0             Divorced    Not-in-family  
48839     HS-grad            9.0              Widowed   Other-relative  
48840   Bachelors           13.0   Married-civ-spouse          Husband  
48841   Bachelors           13.0             Divorced        Own-child  

[48842 rows x 9 columns]

df_W
               Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target  
0                40.0       49084    <=50K  
1                13.0       68553    <=50K  
2                40.0      143654    <=50K  
3                40.0       41510    <=50K  
4                40.0      195025    <=50K  
...               ...         ...      ...  
48838            36.0   735617153   <=50K.  
48839            40.0  1131807446   <=50K.  
48840            50.0   781422087   <=50K.  
48841            40.0   161771296   <=50K.  
48842            60.0  1344995623    >50K.  

[48843 rows x 7 columns]

df_Total
        Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   
48842   NaN      NaN  1344995623      

df_W tiene una row más que df_P: 48843 contra 48842.

In [ ]:
# Acá encontramos que df_W en la columna "ID" tiene el valor 1344995623 extra al df_P:

set(df_W['ID'].unique()) - set(df_P['ID'].unique())

{1344995623}

###1. Cuántos ID tienen en común W y P? Nombrar cuáles no tienen en común:

Tienen en común 48842 IDs. No tienen en común el ID 1344995623.

###2. Corregir los Target a solo dos valores distintos ‘<=50k’ y ‘>50k’

Traemos los valores de la columna Target para ver cuáles debemos corregir:

In [ ]:
df_Total.Target.unique()

array([' <=50K', ' >50K', nan, ' <=50K.', ' >50K.'], dtype=object)

In [ ]:
def corregir_target(x):
   if x in [' <=50K', ' <=50K.']: 
    return '<=50k'
   elif x in [' >50K', ' >50K.']: 
    return '>50k'
   else:
    return x

In [ ]:
df_Total['New_Target'] = df_Total.Target.apply(lambda x: corregir_target(x))

In [ ]:
df_Total['New_Target']= df_Total.Target.transform(lambda x: corregir_target(x) )

df_Total

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   
48842   NaN      NaN  1344995623                  NaN             NaN   

        Education  Education-Num       Marital Status     Relationship  \
0       Bachelors           13.0        Never-married    Not-in-family   
1       Bachelors           13.0   Married-civ-spouse          Husband   
2         HS-grad            9.0             Divorced    Not-in-family   
3            11th            7.0   Married-civ-spouse          Husband   
4       Bachelors           13.0   Married-civ-spouse             Wife   
...           ...            ...                  ...              ...   
48838   Bachelors           13.0             Divorced    Not-in-family   
48839     HS-grad            9.0              Widowed   Other-relative   
48840   Bachelors           13.0   Married-civ-spouse          Husband   
48841   Bachelors           13.0             Divorced        Own-child   
48842         NaN            NaN                  NaN              NaN   

               Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week   Target New_Target  
0                40.0    <=50K      <=50k  
1                13.0    <=50K      <=50k  
2                40.0    <=50K      <=50k  
3                40.0    <=50K      <=50k  
4                40.0    <=50K      <=50k  
...               ...      ...        ...  
48838            36.0   <=50K.      <=50k  
48839            40.0   <=50K.      <=50k  
48840            50.0   <=50K.      <=50k  
48841            40.0   <=50K.      <=50k  
48842            60.0    >50K.       >50k  

[48843 rows x 16 columns]

###Revisamos que hayan quedado correctamente generados los valores únicos solicitados: ‘<=50k’ y ‘>50k’

In [ ]:
df_Total.New_Target.unique()

array(['<=50k', '>50k', nan], dtype=object)

###3. Calcular la proporción que hay entre personas con salarios >50k respecto del total entre personas de distinto Sexo, Raza y Educacion. Por ejemplo cuál es el porcentaje de mujeres con un salario mayor a 50K y compararlo con el de los hombres. Lo mismo para personas de distintas razas y educación


Creamos una nueva columna para contar la cantidad de personas por fila:

In [ ]:
from numpy.ma.core import count
df_Total['Personas'] = df_Total['ID'].apply(lambda x: count(x))
df_Total.head()

Age      Sex      ID    Race         Country   Education  Education-Num  \
0  39.0     Male   49084   White   United-States   Bachelors           13.0   
1  50.0     Male   68553   White   United-States   Bachelors           13.0   
2  38.0     Male  143654   White   United-States     HS-grad            9.0   
3  53.0     Male   41510   Black   United-States        11th            7.0   
4  28.0   Female  195025   Black            Cuba   Bachelors           13.0   

        Marital Status    Relationship          Workclass    fnlwgt  \
0        Never-married   Not-in-family          State-gov   77516.0   
1   Married-civ-spouse         Husband   Self-emp-not-inc   83311.0   
2             Divorced   Not-in-family            Private  215646.0   
3   Married-civ-spouse         Husband            Private  234721.0   
4   Married-civ-spouse            Wife            Private  338409.0   

   Capital Gain  Capital Loss  Hours per week  Target New_Target  Personas  
0        2174.0           0.0            40.0   <=50K      <=50k         1  
1           0.0           0.0            13.0   <=50K      <=50k         1  
2           0.0           0.0            40.0   <=50K      <=50k         1  
3           0.0           0.0            40.0   <=50K      <=50k         1  
4           0.0           0.0            40.0   <=50K      <=50k         1

Hacemos un Groupby para armar la tabla comparativa por Sexo:

In [ ]:
df_group = df_Total.groupby(['Sex', 'New_Target']).agg(sum_personas=('Personas', 'sum'))
df_group

sum_personas
Sex     New_Target              
 Female <=50k              14423
        >50k                1769
 Male   <=50k              22732
        >50k                9917

Hacemos la tabla para el cálculo del promedio:

In [ ]:
df_group['sum_personas'].groupby(level=0).apply(lambda x : (x / float(x.sum())))

Sex      New_Target
 Female  <=50k         0.890749
         >50k          0.109251
 Male    <=50k         0.696254
         >50k          0.303746
Name: sum_personas, dtype: float64

####Repetimos ambos procedimientos para Raza

Hacemos un Groupby para armar la tabla comparativa por Raza:

In [ ]:
df_group = df_Total.groupby(['Race', 'New_Target']).agg(sum_personas=('Personas', 'sum'))
df_group

sum_personas
Race                New_Target              
 Amer-Indian-Eskimo <=50k                415
                    >50k                  55
 Asian-Pac-Islander <=50k               1110
                    >50k                 409
 Black              <=50k               4119
                    >50k                 566
 Other              <=50k                356
                    >50k                  50
 White              <=50k              31155
                    >50k               10606

Hacemos la tabla para el cálculo del promedio:

In [ ]:
df_group['sum_personas'].groupby(level=0).apply(lambda x : (x / float(x.sum())))

Race                 New_Target
 Amer-Indian-Eskimo  <=50k         0.882979
                     >50k          0.117021
 Asian-Pac-Islander  <=50k         0.730744
                     >50k          0.269256
 Black               <=50k         0.879189
                     >50k          0.120811
 Other               <=50k         0.876847
                     >50k          0.123153
 White               <=50k         0.746031
                     >50k          0.253969
Name: sum_personas, dtype: float64

####Repetimos ambos procedimientos para Educación

Hacemos un Groupby para armar la tabla comparativa por Educación:

In [ ]:
df_group = df_Total.groupby(['Education', 'New_Target']).agg(sum_personas=('Personas', 'sum'))
df_group

sum_personas
Education     New_Target              
 10th         <=50k               1302
              >50k                  87
 11th         <=50k               1720
              >50k                  92
 12th         <=50k                609
              >50k                  48
 1st-4th      <=50k                239
              >50k                   8
 5th-6th      <=50k                482
              >50k                  27
 7th-8th      <=50k                893
              >50k                  62
 9th          <=50k                715
              >50k                  41
 Assoc-acdm   <=50k               1188
              >50k                 413
 Assoc-voc    <=50k               1539
              >50k                 522
 Bachelors    <=50k               4712
              >50k                3312
 Doctorate    <=50k                163
              >50k                 431
 HS-grad      <=50k              13281
              >50k                2503
 Masters      <=50k               1198
              >50k                1459
 Preschool    <=50k                 82
              >50k                   1
 Prof-school  <=50k                217
              >50k                 617
 Some-college <=50k               8815
              >50k                2063

Hacemos la tabla para el cálculo del promedio:

In [ ]:
df_group['sum_personas'].groupby(level=0).apply(lambda x : (x / float(x.sum())))

Education      New_Target
 10th          <=50k         0.937365
               >50k          0.062635
 11th          <=50k         0.949227
               >50k          0.050773
 12th          <=50k         0.926941
               >50k          0.073059
 1st-4th       <=50k         0.967611
               >50k          0.032389
 5th-6th       <=50k         0.946955
               >50k          0.053045
 7th-8th       <=50k         0.935079
               >50k          0.064921
 9th           <=50k         0.945767
               >50k          0.054233
 Assoc-acdm    <=50k         0.742036
               >50k          0.257964
 Assoc-voc     <=50k         0.746725
               >50k          0.253275
 Bachelors     <=50k         0.587238
               >50k          0.412762
 Doctorate     <=50k         0.274411
               >50k          0.725589
 HS-grad       <=50k         0.841422
               >50k          0.158578
 Masters       <=50k         0.450884
               >50k     

Las mujeres que cobran por encima de los 50K son solo el 11%, mientras que los hombres que cobran esos salarios alcanzan el 30%.

Las personas del Pacífico Asiático gananan más en promedio que las del resto de las razas, llegando a un 27% de personas con salarios superiores a 50K, mientras que la raza que le sigue es la blanca (25%), y el resto promedia el 12%.

Con respecto al nivel de Educación, observamos que a mayor avance en las carreras, mayores son los salarios que superan los 50K, siendo los de las personas con doctorados los que alcanzan el mayor porcentaje: 72%

Fin

####Todos juntos:

In [ ]:
df_group = df_Total.groupby(['Sex', 'Education', 'Race', 'New_Target']).agg(sum_personas=('Personas', 'sum')).reset_index()

pd.set_option('display.max_rows',500)
pd.set_option('display.max_columns',20)
df_group.head(500)

Sex      Education                 Race New_Target  sum_personas
0     Female           10th   Amer-Indian-Eskimo      <=50k            10
1     Female           10th   Asian-Pac-Islander      <=50k             6
2     Female           10th   Asian-Pac-Islander       >50k             1
3     Female           10th                Black      <=50k            87
4     Female           10th                Other      <=50k             6
5     Female           10th                White      <=50k           344
6     Female           10th                White       >50k             3
7     Female           11th   Amer-Indian-Eskimo      <=50k             7
8     Female           11th   Asian-Pac-Islander      <=50k            12
9     Female           11th                Black      <=50k           124
10    Female           11th                Black       >50k             3
11    Female           11th                Other      <=50k            10
12    Female           11th                White      <=50k           486
13    Female           11th                White       >50k             8
14    Female           12th   Asian-Pac-Islander      <=50k             6
15    Female           12th                Black      <=50k            39
16    Female           12th                Black       >50k             2
17    Female           12th                Other      <=50k             5
18    Female           12th                White      <=50k           157
19    Female           12th                White       >50k             2
20    Female        1st-4th   Amer-Indian-Eskimo      <=50k             2
21    Female        1st-4th   Asian-Pac-Islander      <=50k             4
22    Female        1st-4th                Black      <=50k            10
23    Female        1st-4th                Other      <=50k             3
24    Female        1st-4th                White      <=50k            42
25    Female        5th-6th   Asian-Pac-Islander      <=50k            12
26    Female        5th-6th   Asian-Pac-Islander       >50k             3
27    Female        5th-6th                Black      <=50k            12
28    Female        5th-6th                Other      <=50k             6
29    Female        5th-6th                White      <=50k            93
30    Female        5th-6th                White       >50k             1
31    Female        7th-8th   Amer-Indian-Eskimo      <=50k             3
32    Female        7th-8th   Asian-Pac-Islander      <=50k             5
33    Female        7th-8th                Black      <=50k            39
34    Female        7th-8th                Other      <=50k            10
35    Female        7th-8th                White      <=50k           181
36    Female        7th-8th                White       >50k             1
37    Female            9th   Amer-Indian-Eskimo      <=50k             3
38    Female            9th   Asian-Pac-Islander      <=50k             5
39    Female            9th   Asian-Pac-Islander       >50k             1
40    Female            9th                Black      <=50k            42
41    Female            9th                Other      <=50k             6
42    Female            9th                White      <=50k           159
43    Female            9th                White       >50k             4
44    Female     Assoc-acdm   Amer-Indian-Eskimo      <=50k             7
45    Female     Assoc-acdm   Amer-Indian-Eskimo       >50k             1
46    Female     Assoc-acdm   Asian-Pac-Islander      <=50k            16
47    Female     Assoc-acdm   Asian-Pac-Islander       >50k             5
48    Female     Assoc-acdm                Black      <=50k            89
49    Female     Assoc-acdm                Black       >50k             6
50    Female     Assoc-acdm                Other      <=50k             2
51    Female     Assoc-acdm                Other       >50k             2
52    Female     Assoc-acdm                White      <=50k           427
53    Female 